fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,479 @@
+        Ok(v)
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmResult<Vec<CrossMsg>, Error> {
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
+                .map_err(Error::NonSequentialParentViewInsert);
+            (cache, r)
+        })?;
+
+        if let Err(e) = r {
+            return abort(e);
+        }
+
+        Ok(())
+    }
+}
+
+impl ParentFinalityProvider for InMemoryFinalityProvider {
+    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
+        let finality = self
+            .last_committed_finality
+            .read_clone()?
+            .ok_or(StmError::Abort(Error::CommittedFinalityNotReady))?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
+        let last_committed_finality = self.last_committed_finality()?;
+
+        // parent height is not ready to be proposed yet
+        if height <= last_committed_finality.height {
+            return Ok(None);
+        }
+
+        // prepare block hash and validator set
+        let height_data = self.parent_view_data.height_data.read()?;
+        let block_hash = if let Some(v) = height_data.get_value(height) {
+            let (block_hash, _, _) = v;
+            block_hash.clone()
+        } else {
+            return abort(Error::HeightNotFoundInCache(height));
+        };
+
+        Ok(Some(IPCParentFinality { height, block_hash }))
+    }
+
+    fn check_proposal(&self, proposal: &IPCParentFinality) -> StmResult<(), Error> {
+        self.check_height(proposal)?;
+        self.check_block_hash(proposal)
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) -> StmResult<(), Error> {
+        // the height to clear
+        let height = finality.height;
+
+        self.parent_view_data.height_data.update(|mut cache| {
+            cache.remove_key_below(height + 1);
```

Comment:
I wonder if this might be working on the previous assumption where the cache contained the blocks all the way to the tip of the chain, rather than just the finalized block. There, if you had the chain cached up to 102, you proposed for finality height 92, and when it was executed you removed everything below 93. The cache still remembered what the highest height was. 

Now, it only contains entries up to 92, so by removing basically it clears out potentially everything. Is that okay?

Reply:
I think it is because we write to `last_committed_finality`, just wanted to make sure there's no problem because you rely so much on insertions range checks.

Reply:
From `sync` point of view, this should be ok. This implies that the parent is producing block slowly or completely stopped and it means there should not be any parent view related proposal.